Docs: fix links to methods with String default values

bcardiff opened the pull request:
The double-quotes value in the id need to be escaped with &quot; in order to work.

For example https://crystal-lang.org/api/0.34.0/Path.html#new(name:String=%22%22):Path-class-method is not working without this PR.

Blacksmoke16 commented:
Fixes #8809.

jhass commented:
Wait, why is there any JS  for this? Browsers should support scrolling to an element with the right `id` natively, as long as we keep consistency in generating links and `id=` tags, it should just work, no?

bcardiff commented:
I would expect that also @jhass . I'm not sure how what other JS are required in the documentation html to expand/hightlight methods.

Maybe @straight-shoota is more aware.

straight-shoota commented:
Seems like this was introduced in #3623 to fix an issue with scroll position being reset when reloading the page. Affecting every fragment link action seems quite intrusive, though.
/cc @Sija

straight-shoota commented:
I actually can't notice any difference in the user experience when this method is removed. #3623 lacks a detailled explanation what exactly goes wrong, so I might be missing something.

Sija commented:
@straight-shoota That was ~3,5yrs ago, I lot could have changed since then (and did), so I'd suggest trying out whether this workaround is still needed and applicable these days.

straight-shoota commented:
Yeah, but I'm not even sure what *exactly* this workaround fixed. Can you recall any specific behaviour we can test?

Sija commented:
@straight-shoota It was related to scrolling - as mentioned in the original PR - this should be enough to  let you test it - both ff and chrome woud be needed as their behavior differs in such matters.

straight-shoota commented:
Okay so I found the faulty behaviour. When you reload the page in Chrome it always scrolls the main content pane to the top, even if the URL points to an anchor. This seems to be related to the script responsible for the sidebar to scroll the currently selected entry into view. That should just affect the sidebar. But Chrome seems to disagree.

But the fix to this problem should be simpler than what we currently do.
```js
var elem = document.getElementById(window.location.hash.substr(1))
if (elem) elem.scrollIntoView();
```
This only needs to be executed once, not on every hashchange event. Klicking anchor links on the page should not be affected. It's only relevant when the page is loaded.

bcardiff commented:
It seems unrelated to this change then.

For ids with quotes  the following is still needed:

```js
document.getElementById(unescape(window.location.hash.substr(1)).replace(/"/g, '&quot;'))
```

Another approach would be to use safe chars for ids.

Either way, a story for another PR it seems.

straight-shoota commented:
No, this is the wrong change. It just fixes a symptom in our JavaScript code. Instead we should fix the cause, that is the anchor names and link targets must work even when special characters are involved.

straight-shoota commented:
It's not even enough because there are other characters that cause issues (for example `|`).
This is the fragment for `File.open` that Firefox produces when clicking on the link in the constructors section:
```
#open(filename:Path%7CString,mode="r",perm=DEFAULT_CREATE_PERMISSIONS,encoding=nil,invalid=nil):self-class-method
```
This fragment actually works
```
#open(filename:Path|String,mode=&quot;r&quot;,perm=DEFAULT_CREATE_PERMISSIONS,encoding=nil,invalid=nil):self-class-method
```

bcardiff commented:
But is wrong because we are mixing how `"` need to be escaped in uri and html.

The ideal is to use safe chars for ids.

straight-shoota commented:
> The ideal is to use safe chars for ids.

Maybe. But the characters arenot really the problem. They actually work fine:

```html
<a href="#&quot;|">goto</a>
<a id="&quot;|">target</a>
```
So the error is on our side how we process the anchor values. For example, id values must not be URI escaped.

bcardiff commented:
Ok, drilled down a little bit, the issue was that the arguments are escaped twice.
I removed the query selector in favor of getElementById
I removed the hashchange event
But I didn't add the `scrollIntoView` because the note at https://caniuse.com/#search=scrollIntoView

Sija commented:
> I removed the hashchange event

Did you tested did the current version works with the _Back_ browser button - i.e. correctly scrolls to the given _id_?

bcardiff commented:
Now I did @Sija . Thanks 🙈

straight-shoota commented:
We need to make add a comment to explain why this code is there and how it works.

bcardiff commented:
> We need to make add a comment to explain why this code is there and how it works.

Definitely, but I only know that is need. I haven't tracked why/how it works.
